src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.LegacyEditor/RazorWrapperFactory.cs

```diff
@@ -0,0 +1,201 @@
+
+    internal static IRazorTagHelperBinding WrapTagHelperBinding(object obj)
+        => Wrap<TagHelperBinding, TagHelperBindingWrapper, IRazorTagHelperBinding>(obj, static obj => new TagHelperBindingWrapper(obj));
+
+    internal static IRazorTagHelperDescriptor WrapTagHelperDescriptor(object obj)
+        => Wrap<TagHelperDescriptor, TagHelperDescriptorWrapper, IRazorTagHelperDescriptor>(obj, static obj => new TagHelperDescriptorWrapper(obj));
+
+    internal static IRazorTagHelperDocumentContext WrapTagHelperDocumentContext(object obj)
+        => Wrap<TagHelperDocumentContext, TagHelperDocumentContextWrapper, IRazorTagHelperDocumentContext>(obj, static obj => new TagHelperDocumentContextWrapper(obj));
+
+    internal static IRazorTagHelperCompletionService WrapTagHelperCompletionService(object obj)
+        => Wrap<TagHelperCompletionService, TagHelperCompletionServiceWrapper, IRazorTagHelperCompletionService>(obj, static obj => new TagHelperCompletionServiceWrapper(obj));
+
+    internal static IRazorTagHelperFactsService WrapTagHelperFactsService(object obj)
+        => Wrap<TagHelperFactsService, TagHelperFactsServiceWrapper, IRazorTagHelperFactsService>(obj, static obj => new TagHelperFactsServiceWrapper(obj));
+
+    internal static IRazorTagMatchingRuleDescriptor WrapTagMatchingRuleDescriptor(object obj)
+        => Wrap<TagMatchingRuleDescriptor, TagMatchingRuleDescriptorWrapper, IRazorTagMatchingRuleDescriptor>(obj, static obj => new TagMatchingRuleDescriptorWrapper(obj));
+
+    internal static RazorSourceSpan ConvertSourceSpan(object obj)
+    {
+        var span = CastStruct<SourceSpan>(obj);
+
+        return ConvertSourceSpan(span);
+    }
+
+    [return: NotNullIfNotNull(nameof(obj))]
+    internal static RazorSourceChange? ConvertSourceChange(object? obj)
+    {
+        return CastObject<SourceChange>(obj) is SourceChange change
+            ? new RazorSourceChange(
+                Span: ConvertSourceSpan(change.Span),
+                NewText: change.NewText)
+            : null;
+    }
+
+    [return: NotNullIfNotNull(nameof(obj))]
+    internal static RazorSourceMapping? ConvertSourceMapping(object? obj)
+    {
+        return CastObject<SourceMapping>(obj) is SourceMapping mapping
+            ? new RazorSourceMapping(
+                OriginalSpan: ConvertSourceSpan(mapping.OriginalSpan),
+                GeneratedSpan: ConvertSourceSpan(mapping.GeneratedSpan))
+            : null;
+    }
+
+    private static RazorSourceSpan ConvertSourceSpan(SourceSpan span)
+        => new(
+            FilePath: span.FilePath,
+            AbsoluteIndex: span.AbsoluteIndex,
+            LineIndex: span.LineIndex,
+            CharacterIndex: span.CharacterIndex,
+            Length: span.Length,
+            LineCount: span.LineCount,
+            EndCharacterIndex: span.EndCharacterIndex);
+
+    [return: NotNullIfNotNull(nameof(obj))]
+    private static T? CastObject<T>(object? obj)
```

Comment:
If we get rid of the object based params, then I don't think we need the `Cast{Object,Struct}` methods?

Reply:
Unfortunately, we can't get rid of those if we want to hide the compiler types from the legacy editor.